Development and validation of a SNP genotyping panel for the
differentiation of Cannabis sativa chemovars
================

## Project description

This repository hosts the code used to develop a SNP-based genotyping
panel. This pipeline combines the work of many others. The primary
objective of sharing this code is to further document the approach we
used to develop our genotyping panel, largely in an effort to facilitate
the task of others who are seeking to keep track of their cannabis
genetics using PCR-based assays. We suggest that those seeking greater
detail consult the associate article, *Development and validation of SNP
genotyping panel for the differentiation of Cannabis sativa chemovars*,
available [here](link).


## Bibliography

Danecek, P., Auton, A., Abecasis, G., Albers, C.A., Banks, E., DePristo,
M.A., Handsaker, R.E., Lunter, G., Marth, G.T., Sherry, S.T., McVean,
G., Durbin, R., 1000 Genomes Project Analysis Group, 2011. The variant
call format and VCFtools. Bioinformatics 27, 2156–2158.
<https://doi.org/10.1093/bioinformatics/btr330>

Langmead, B., Salzberg, S.L., 2012. Fast gapped-read alignment with
Bowtie 2. Nat. Methods 9, 357–359. <https://doi.org/10.1038/nmeth.1923>

Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N.,
Marth, G., Abecasis, G., Durbin, R., 1000 Genome Project Data Processing
Subgroup, 2009. The Sequence Alignment/Map format and SAMtools.
Bioinforma. Oxf. Engl. 25, 2078–2079.
<https://doi.org/10.1093/bioinformatics/btp352>

McKinney, W., 2010. Data Structures for Statistical Computing in Python.
Presented at the Python in Science Conference, Austin, Texas, pp. 56–61.
<https://doi.org/10.25080/Majora-92bf1922-00a>

Quinlan, A.R., Hall, I.M., 2010. BEDTools: a flexible suite of utilities
for comparing genomic features. Bioinformatics 26, 841–842.
<https://doi.org/10.1093/bioinformatics/btq033>
